src/main/java/com/iota/iri/service/transactionpruning/async/MilestonePrunerJobQueue.java

```diff
@@ -0,0 +1,162 @@
+     */
+    public MilestonePrunerJobQueue(TransactionPruner transactionPruner, SnapshotConfig snapshotConfig) {
+        this.transactionPruner = transactionPruner;
+
+        youngestFullyCleanedMilestoneIndex = snapshotConfig.getMilestoneStartIndex();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This queue will only accept {@link MilestonePrunerJob}s, since it is tailored to the logic related to cleaning up
+     * milestones.
+     *
+     * Since the {@link MilestonePrunerJob}s can take a long time to finish, we try to consolidate the queue when we add
+     * new jobs, so the state file doesn't get to big while the node is busy cleaning up the milestones. To do so, we
+     * first check if the job that shall be added isn't covered by existing cleanup jobs, yet. If it targets a milestone
+     * outside of the already covered range, we also check if the last existing job can be extended to cover the target
+     * milestone index of our new job.
+     *
+     * If the job can not be appended to an existing job, we add it to the end of our queue.
+     *
+     * @param job the {@link MilestonePrunerJob} that shall be added to the queue
+     * @throws TransactionPruningException if the given job is no {@link MilestonePrunerJob}
+     */
+    @Override
+    public void addJob(TransactionPrunerJob job) throws TransactionPruningException {
+        if (!(job instanceof MilestonePrunerJob)) {
+            throw new TransactionPruningException("the MilestonePrunerJobQueue only supports MilestonePrunerJobs");
+        }
+
+        synchronized (jobs) {
+            MilestonePrunerJob newMilestonePrunerJob = (MilestonePrunerJob) job;
+            MilestonePrunerJob lastMilestonePrunerJob = (MilestonePrunerJob) jobs.peekLast();
+
+            // determine where the last job stops / stopped cleaning up
+            int lastTargetIndex = lastMilestonePrunerJob != null
+                                ? lastMilestonePrunerJob.getTargetIndex()
+                                : youngestFullyCleanedMilestoneIndex;
+
+            // if the cleanup target of our job is covered already -> don't add
+            if (newMilestonePrunerJob.getTargetIndex() <= lastTargetIndex) {
+                return;
+            }
+
+            // adjust the job to reflect the progress that was / will be made by previous jobs in the queue
+            if (lastTargetIndex >= newMilestonePrunerJob.getStartingIndex()) {
+                newMilestonePrunerJob.setStartingIndex(lastTargetIndex + 1);
+            }
+            if (lastTargetIndex >= newMilestonePrunerJob.getCurrentIndex()) {
+                newMilestonePrunerJob.setCurrentIndex(lastTargetIndex + 1);
+            }
+
+            // if the last job is not DONE, we can just extend it to also clean our new target
+            if (lastMilestonePrunerJob != null) {
+                synchronized (lastMilestonePrunerJob) {
```

Comment:
because the job might just have finished and left the loop already that keeps it processing but not updated the status yet, thats why i synchronize the status change over the job itself. See comment in MilestonePrunerJob class